redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,709 @@
+    /// This method iterates over the scanned slots map and finds the next slot that is set to 0.
+    ///
+    /// # Arguments
+    ///
+    /// * `scanned_slots_map` - The scanned slots map.
+    ///
+    /// # Returns
+    ///
+    /// The next slot that hasn't been scanned, or `None` if all slots have been scanned.
+    pub fn get_next_slot(&self, scanned_slots_map: &SlotsBitsArray) -> Option<u16> {
+        let all_slots_scanned = scanned_slots_map.iter().all(|&word| word == u64::MAX);
+        if all_slots_scanned {
+            return Some(END_OF_SCAN);
+        }
+        for (i, slot) in scanned_slots_map.iter().enumerate() {
+            let mut mask = 1;
+            for j in 0..BITS_PER_U64 {
+                if (slot & mask) == 0 {
+                    return Some((i * BITS_PER_U64 + j) as u16);
+                }
+                mask <<= 1;
+            }
+        }
+        None
+    }
+
+    /// Update the next address to be scanned without updating the scanned slots map.
+    /// This method is used when the address epoch has changed and we can't know which slots are new or when slots are missing.
+    /// In this case we will skip updating the scanned_slots_map and will just update the address and the cursor.
+    ///
+    /// # Arguments
+    ///
+    /// * `connection` - The connection to the Redis cluster.
+    ///
+    /// # Returns
+    ///
+    /// The updated ScanState.
+    pub(crate) async fn update_scan_state_without_updating_scanned_map<
+        C: ClusterInScan + ?Sized,
+    >(
+        &self,
+        connection: &C,
+    ) -> RedisResult<ScanState> {
+        let next_slot = self.get_next_slot(&self.scanned_slots_map).unwrap();
+        let new_address = if next_slot == END_OF_SCAN {
+            Ok("".to_string())
+        } else {
+            connection.get_address_by_slot(next_slot).await
+        };
+        match new_address {
+            Ok(address) => {
+                let new_epoch = connection.get_address_epoch(&address).await.unwrap_or(0);
+                let new_cursor = 0;
+                Ok(ScanState::create(
+                    new_cursor,
+                    self.scanned_slots_map,
+                    address,
+                    new_epoch,
+                ))
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Update the next address to be scanned.
+    ///
+    /// This method updates the scan state by updating the scanned slots map, determining the next slot to be scanned,
+    /// retrieving the address for that slot and the epoch of that address, and creating a new ScanState.
+    ///
+    /// # Arguments
+    ///
+    /// * `connection` - The connection to the Redis cluster.
+    ///
+    /// # Returns
+    ///
+    /// The updated ScanState.
+    async fn update_scan_state_and_get_next_address<C: ClusterInScan + ?Sized>(
+        &mut self,
+        connection: &C,
+    ) -> RedisResult<ScanState> {
+        let _ = connection.refresh_slots().await;
```

Comment:
We have no way to promise a full scan without it.
While epoch was the first solution, apparently epoch is changing for fail-over or for slots added.
When slots are moving out, the epoch doesn't change.
So if the epoch didn't change, and we want to know which slots we scanned, we have to check which slots are still there at the end of the scan, since nothing will indicate if slots moved out of the node. A full scan of a cluster, even big take less than a minute, so we cant even trust the topology refresh happen every 60 sec.
Thats why we dont check every scan as we design it, but every node, since it become expensive.

Reply:
then you'll need to clearly document this to the user - if using scan ocassionally blocks the connection from receiving new commands, this needs to be made explicit, otherwise you'll get angry users who don't understand why they see TPS drops during scans. Did you talk to users, and check that they prefer slow scans over almost-perfect scans which are much more performant?

You said that ATM a full scan could take a minute, which is an eternity in a sub-millisecond world. How long will it take if you didn't referesh slots after every node?

Reply:
Take less than a minute.
What is the goal of all this complicated code if at the end of the day we cant promise thats its a full scan?
We can offer if it will be requested a `almost_perfect_cluster_scan` which will be not checking anything but scan node by node since the case of slot migration during scan is anyway rare.
Cluster slots is not a blocking command, and i dont think it is such a price if you use async_client, you still can work and the server is not block. it does put weight on tcp connection.
i think at this point we will not change the abstract idea, scan is not a command that is usually checked for performance, and even tho we should check for customers opinions, since we currently dont have the i will go with mentioning it in the docs and later on may change if requested.    
Probably add an additional option than change the current.

Reply:
> Take less than a minute.

again - when you're usually talking in the order of magnitude of milliseconds, even talking in the order of magnitude of 10s of seconds is REALLY bad. 

> We can offer if it will be requested a almost_perfect_cluster_scan which will be not checking anything but scan node by node since the case of slot migration during scan is anyway rare.

https://github.com/amazon-contributing/redis-rs/pull/141#discussion_r1647649327
There you're arguing for simplicity and failing in the rare cases. Here you're arguing for complexity and covering the rare cases. You're arguing a point here, and its opposite somewhere else - this isn't a coherent design.
The scan you've implemented isn't perfect, and that's ok - not only that, since you're running in the order of magnitude of a minute, the scan truly isn't perfect, because the cluster could undergo a lot of changes in this time. While you're technically upholding the interface of "reporting slots as they are at the time", you're not in the spirit of "returning the whole state of the server in small chunks".

Reply:
in any case, if the design is finalized and you don't intend to change it, there's no need to continue this conversation.

Reply:
A simpler solution is to just call CLUSTER SlOTS on the read node, which will tell you whether the node things there was a failover, and will save the overhead of sending to multiple nodes and waiting for all operations to complete in order to take a write lock on the connection's internal state.

Reply:
> since you're running in the order of magnitude of a minute, the scan truly isn't perfect, because the cluster could undergo a lot of changes in this time. While you're technically upholding the interface of "reporting slots as they are at the time", you're not in the spirit of "returning the whole state of the server in small chunks".
Im not sure i got your point.
it is actually magnitude of ms, a cluster with 45 nodes and 100000 keys will take around 600 ms, which is not so much and its count as pretty big cluster.

> A simpler solution is to just call CLUSTER SlOTS on the read node, which will tell you whether the node things there was a failover, and will save the overhead of sending to multiple nodes and waiting for all operations to complete in order to take a write lock on the connection's internal state.
That's sound like a good solution but it's create a lot of extra complexity to the code, since we need to track the state of the whole cluster, and compare it in order to know if it changed and to refresh topology for it.

> returning the whole state of the server in small chunks
This is not the spirit of scan - the spirit is to return in small chunks to the parts of the state that stayed the same from start to the end.

I agree that the scan is not perfect, and you can always sharpen it, but at this point we need to publish a sufficient method (GA is coming) and later, after GA, make it better. When the knife is not on our neck 🥵

That's why at this point we don't i will not make changes to the method itself but ill fix holes and style.